File: services/orders/src/Orders.Infrastructure/Id/RequestId.cs
using System.Text.RegularExpressions;

namespace Orders.Infrastructure.Id;

/// <summary>
/// The cross-service correlation id — <c>req_</c> + a nano-id of
/// <see cref="NanoIdConfig.Length"/> characters — carried on every log line of a request and
/// forwarded to every downstream hop. See [[nano-id]]
/// </summary>
public static class RequestId
{
    /// <summary>
    /// The header the id travels in, inbound and outbound. Lowercase, matching every other
    /// service here; ASP.NET Core's lookup is case-insensitive.
    /// </summary>
    public const string HeaderName = "x-request-id";

    /// <summary>The <c>prefix_nanoid</c> prefix for a request id — see [[nano-id]].</summary>
    public const string Prefix = NanoIdConfig.RequestPrefix;

    /// <summary>
    /// <c>req_</c> plus <see cref="NanoIdConfig.Length"/> characters of the alphabet.
    /// CONTRACT: Derive the pattern from <see cref="NanoIdConfig"/>, never write it out — a
    /// copy drifts the day the format changes and then rejects every id the generator
    /// produces. See [[nano-id]]
    /// </summary>
    private static readonly Regex Pattern = new(
        NanoIdConfig.PatternFor(Prefix),
        RegexOptions.Compiled | RegexOptions.CultureInvariant);

    /// <summary>A new correlation id.</summary>
    public static string New() => NanoId.NewId(Prefix);

    /// <summary>
    /// The caller's request id if it is one of ours, otherwise a freshly minted one.
    /// CONTRACT: Validate the header — untrusted, yet stamped onto every log line and
    /// forwarded to Tracking and SQS. An unbounded string bloats every record and an
    /// <c>ord_</c>-shaped value makes log queries correlate the wrong things.
    /// CONTRACT: Do NOT answer 400 on a bad one — a mangling proxy would turn an
    /// observability nicety into an outage. See [[logging-context]]
    /// </summary>
    public static string Resolve(string? headerValue) =>
        headerValue is not null && Pattern.IsMatch(headerValue) ? headerValue : New();
}

File: services/users/src/shared/logging/request-id.ts
import { NanoIdConfig } from "#shared/id/nano-id";

/**
 * The header carrying the correlation id between services.
 *
 * Lowercase because Node normalises incoming header names; outbound callers in
 * this repo send the same spelling.
 */
export const REQUEST_ID_HEADER = "x-request-id";

/**
 * The `prefix_nanoid` prefix for a request id, read from the shared config
 * rather than declared here — see [[nano-id]]. Every prefix this service mints
 * lives in that one map so they can be audited for collisions together.
 */
const REQUEST_ID_PREFIX = NanoIdConfig.PREFIXES.Request;

/**
 * `req_` plus the shared nano-id alphabet and length, DERIVED from NanoIdConfig rather
 * than written out — a hand-written pattern is how a service starts rejecting its own
 * ids after the format changes.
 */
const REQUEST_ID_PATTERN = NanoIdConfig.pattern(REQUEST_ID_PREFIX);

/** A new correlation id. */
export function generateRequestId(): string {
  return NanoIdConfig.newRequestId();
}

/**
 * The request id for an inbound request: the caller's if it is one of ours,
 * otherwise a fresh one.
 *
 * CONTRACT: Validate the header, and discard silently rather than rejecting the
 * request. It is untrusted input copied onto EVERY log line of the flow and forwarded
 * downstream, so an unbounded value bloats the stream, control characters corrupt the
 * JSON a dashboard parses, and another service's id shape makes a query correlate the
 * wrong things. A 400 would turn an observability nicety into an outage; a fresh id
 * keeps the flow correlated end to end.
 * See [[2026-08-15-request-id-correlation-design]]
 */
export function resolveRequestId(headerValue: unknown): string {
  return typeof headerValue === "string" && REQUEST_ID_PATTERN.test(headerValue)
    ? headerValue
    : generateRequestId();
}
